x-pack/test_serverless/api_integration/services/svl_reporting.ts

```diff
@@ -0,0 +1,182 @@
+      });
+    },
+
+    /**
+     * Define a role that DOES grant privileges to create certain types of reports.
+     */
+    async createReportingRole() {
+      await security.role.create(REPORTING_ROLE, {
+        metadata: {},
+        elasticsearch: {
+          cluster: [],
+          indices: [
+            {
+              names: ['ecommerce'],
+              privileges: ['read', 'view_index_metadata'],
+              allow_restricted_indices: false,
+            },
+          ],
+          run_as: [],
+        },
+        kibana: [
+          {
+            base: [],
+            feature: { discover: ['minimal_read', 'generate_report'] },
+            spaces: ['*'],
+          },
+        ],
+      });
+    },
+
+    async createReportingUser(
+      username = REPORTING_USER_USERNAME,
+      password = REPORTING_USER_PASSWORD
+    ) {
+      await security.user.create(username, {
+        password,
+        roles: [REPORTING_ROLE],
+        full_name: 'Reporting User',
+      });
+    },
+
+    /**
+     * Use the internal API to create any kind of report job
+     */
+    async createReportJobInternal(
+      jobType: string,
+      job: object,
+      username: string,
+      password: string
+    ) {
+      const requestPath = `${INTERNAL_ROUTES.GENERATE_PREFIX}/${jobType}`;
+      log.debug(`POST request to ${requestPath}`);
+
+      const { status, body } = await supertest
+        .post(requestPath)
+        .auth(username, password)
+        .set(...API_HEADER)
+        .set(...INTERNAL_HEADER)
+        .send({ jobParams: rison.encode(job) });
+
+      expect(status).to.be(200);
+
+      return {
+        job: (body as ReportingJobResponse).job,
+        path: (body as ReportingJobResponse).path,
+      };
+    },
+
+    async waitForJobToFinish(
+      downloadReportPath: string,
+      username: string,
+      password: string,
+      options?: { timeout?: number }
+    ) {
+      await retry.waitForWithTimeout(
+        `job ${downloadReportPath} finished`,
+        options?.timeout ?? config.get('timeouts.kibanaReportCompletion'),
+        async () => {
+          const response = await supertest
+            .get(`${downloadReportPath}?elasticInternalOrigin=true`)
+            .auth(username, password)
+            .responseType('blob')
+            .set(...API_HEADER)
+            .set(...INTERNAL_HEADER);
+
+          if (response.status === 503) {
+            log.debug(`Report at path ${downloadReportPath} is pending`);
+            return false;
+          }
+
+          log.debug(`Report at path ${downloadReportPath} returned code ${response.status}`);
+
+          if (response.status === 200) {
+            log.debug(`Report at path ${downloadReportPath} is complete`);
+            return true;
+          }
+
+          throw new Error(`unexpected status code ${response.status}`);
+        }
+      );
+    },
+
+    async getCompletedJobOutput(downloadReportPath: string, username: string, password: string) {
```

Comment:
We should add a comment about only using this function within an API test. In a functional test, we have to click the download link in the UI.